cmd/config/docs/api-conventions/functions-spec.md

```diff
@@ -18,168 +23,372 @@ interpreted as described in [RFC 2119][2].
+    file:
+      path: service.yaml
+```
 
 ### Serialization Format
 
 A function MUST support YAML as a serialization format for the input and output.
-A function MUST use utf8 encoding (as YAML is a superset of JSON, JSON will also be supported
-by any conforming function).
+A function MUST use utf8 encoding (as YAML is a superset of JSON, JSON will also
+be supported by any conforming function).
+
+### Containerization
+
+A function MUST be implemented as a container.
+
+A function container MUST be capable of running as a non-root user `nobody` if
+it does not require access to host filesystem or makes network calls.
+
+### stderr
+
+Any non-structured error messages MUST be emitted to `stderr`. `stdout` is
+reserved for `ResourceList` as described above.
+
+### Exit Code
+
+An exit code of zero indicates function execution was successful. A non-zero
+exit code indicates a failure.
 
 ### Operations
 
-A function MAY Create, Update, or Delete any number of items in the `items` field and output the
-resultant list.
+A function MAY Create, Update, or Delete any number of items in the `items`
+field and output the resultant list in the corresponding `items` field of the
+output.
 
-A function MAY modify annotations with prefix `config.kubernetes.io`, but must be careful about
-doing so since they’re used for orchestration purposes and will likely impact subsequent functions
-in the pipeline.
+A function MAY modify annotations with prefix `config.kubernetes.io`, but must
+be careful about doing so since they’re used for orchestration purposes and will
+likely impact subsequent functions in the pipeline. Several annotations and
+their semantics are listed below.
 
 A function SHOULD preserve comments when input serialization format is YAML.
-This allows for human authoring of configuration to coexist with changes made by functions.
+This allows for human authoring of configuration to coexist with changes made by
+functions.
 
-### Containerization
+### Annotations
 
-A function MUST be implemented as a container.
+The orchestrator and function can communicate on a per-resource basis using the
+following annotations:
+
+#### `config.kubernetes.io/path`
+
+Records the slash-delimited, OS-agnostic, relative file path to a resource. The
+path is relative to a fix location on the filesystem. Different orchestrator
+implementations can choose different fixed points.
+
+This annotation only exists in the wire format and is not persisted to the
+filesystem. The orchestrator sets this annotation when reading files from the
+local filesystem and removes the annotation when writing the output of functions
+back to the filesystem.
 
-A function container MUST be capable of running as a non-root user if it does not require
-access to host filesystem or makes network calls.
+Example:
 
-### stdin/stdout/stderr and Exit Codes
+```yaml
+metadata:
+  annotations:
+    config.kubernetes.io/path: "relative/file/path.yaml"
+```
+
+#### `config.kubernetes.io/index`
```

Comment:
I have seen `config.k8s.io/id` been set in a resourceList. Is it an annotation that we want documented or is it being deprecated?

Reply:
`config.k8s.io/id` is a universal ID in all input resources. I think yes we should document it.

Reply:
I changed the wording so that the function MUST NOT modify `config.kubernetes.io` annotations except `index` and `path` which it SHOULD NOT modify:

This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.